File: ios-mma-recomp/MMARecomp/Models/WeighIn.swift
import Foundation

/// Flags contextuels qui faussent le %BF / le poids brut — affichés en note
/// à côté du point plutôt que cachés, pour désamorcer les fausses alertes.
struct WeighInContext: Codable, Hashable {
    var creatineRecente: Bool = false
    var alcoolRecent: Bool = false
    var postTraining: Bool = false

    enum CodingKeys: String, CodingKey {
        case creatineRecente = "creatine_recente"
        case alcoolRecent = "alcool_recent"
        case postTraining = "post_training"
    }

    var hasAnyFlag: Bool { creatineRecente || alcoolRecent || postTraining }

    var flagLabels: [String] {
        var labels: [String] = []
        if creatineRecente { labels.append("créatine récente") }
        if alcoolRecent { labels.append("alcool récent") }
        if postTraining { labels.append("post-training") }
        return labels
    }
}

struct WeighIn: Codable, Identifiable {
    var id: UUID
    var userId: UUID
    var date: String
    var heure: String // "HH:mm:ss"
    var type: WeighInType
    var poidsKg: Double
    var bfPct: Double?
    var contexte: WeighInContext
    var createdAt: String?

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case date, heure, type
        case poidsKg = "poids_kg"
        case bfPct = "bf_pct"
        case contexte
        case createdAt = "created_at"
    }
}

struct NewWeighIn: Codable {
    var date: String
    var heure: String
    var type: WeighInType
    var poidsKg: Double
    var bfPct: Double?
    var contexte: WeighInContext

    enum CodingKeys: String, CodingKey {
        case date, heure, type
        case poidsKg = "poids_kg"
        case bfPct = "bf_pct"
        case contexte
    }
}

File: android-mma-recomp/app/src/main/java/com/example/mmarecomp/model/WeighIn.kt
package com.example.mmarecomp.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** Flags contextuels qui faussent le %BF / le poids brut — affichés en note
 *  à côté du point plutôt que cachés, pour désamorcer les fausses alertes. */
@Serializable
data class WeighInContext(
    @SerialName("creatine_recente") val creatineRecente: Boolean = false,
    @SerialName("alcool_recent") val alcoolRecent: Boolean = false,
    @SerialName("post_training") val postTraining: Boolean = false,
) {
    val hasAnyFlag: Boolean get() = creatineRecente || alcoolRecent || postTraining

    val flagLabels: List<String>
        get() = buildList {
            if (creatineRecente) add("créatine récente")
            if (alcoolRecent) add("alcool récent")
            if (postTraining) add("post-training")
        }
}

@Serializable
data class WeighIn(
    val id: String,
    @SerialName("user_id") val userId: String,
    val date: String,
    val heure: String, // "HH:mm:ss"
    val type: WeighInType,
    @SerialName("poids_kg") val poidsKg: Double,
    @SerialName("bf_pct") val bfPct: Double? = null,
    val contexte: WeighInContext = WeighInContext(),
    @SerialName("created_at") val createdAt: String? = null,
)

@Serializable
data class NewWeighIn(
    val date: String,
    val heure: String,
    val type: WeighInType,
    @SerialName("poids_kg") val poidsKg: Double,
    @SerialName("bf_pct") val bfPct: Double? = null,
    val contexte: WeighInContext,
)
